src/magic_enchantment.cpp

```diff
@@ -0,0 +1,453 @@
+            if( mult != 0.0 ) {
+                values_multiply.emplace( value, mult );
+            }
+        }
+    }
+}
+
+void enchantment::serialize( JsonOut &json ) const
+{
+    json.start_object();
+
+    json.member( "has", io::enum_to_string<has>( active_conditions.first ) );
+    json.member( "condition", io::enum_to_string<condition>( active_conditions.second ) );
+
+    if( !hit_you_effect.empty() ) {
+        json.member( "hit_you_effect" );
+        json.start_array();
+        for( const fake_spell &sp : hit_you_effect ) {
+            sp.serialize( json );
+        }
+        json.end_array();
+    }
+
+    if( !hit_me_effect.empty() ) {
+        json.member( "hit_me_effect" );
+        json.start_array();
+        for( const fake_spell &sp : hit_me_effect ) {
+            sp.serialize( json );
+        }
+        json.end_array();
+    }
+
+    if( !intermittent_activation.empty() ) {
+        json.member( "intermittent_activation" );
+        json.start_object();
+        for( const std::pair<time_duration, std::vector<fake_spell>> pair : intermittent_activation ) {
+            json.member( "duration", pair.first );
+            json.start_array( "effects" );
+            for( const fake_spell &sp : pair.second ) {
+                sp.serialize( json );
+            }
+            json.end_array();
+        }
+        json.end_object();
+    }
+
+    json.member( "values" );
+    json.start_array();
+    for( int value = 0; value < mod::NUM_MOD; value++ ) {
+        mod enum_value = static_cast<mod>( value );
+        if( get_value_add( enum_value ) == 0 && get_value_multiply( enum_value ) == 0.0 ) {
+            continue;
+        }
+        json.start_object();
+        json.member( "value", io::enum_to_string<mod>( enum_value ) );
+        if( get_value_add( enum_value ) != 0 ) {
+            json.member( "add", get_value_add( enum_value ) );
+        }
+        if( get_value_multiply( enum_value ) != 0 ) {
+            json.member( "multiply", get_value_multiply( enum_value ) );
+        }
+        json.end_object();
+    }
+    json.end_array();
+
+    json.end_object();
+}
+
+bool enchantment::stacks_with( const enchantment &rhs ) const
+{
+    return active_conditions == rhs.active_conditions;
+}
+
+bool enchantment::add( const enchantment &rhs )
+{
+    if( !stacks_with( rhs ) ) {
+        return false;
+    }
+    for( const std::pair<mod, int> &pair_values : rhs.values_add ) {
+        values_add[pair_values.first] += pair_values.second;
+    }
+    for( const std::pair<mod, double> &pair_values : rhs.values_multiply ) {
+        // values do not multiply against each other, they add.
+        // so +10% and -10% will add to 0%
+        values_multiply[pair_values.first] += pair_values.second;
+    }
+    for( const fake_spell &fake : rhs.hit_me_effect ) {
+        hit_me_effect.emplace_back( fake );
+    }
+    for( const fake_spell &fake : rhs.hit_you_effect ) {
+        hit_you_effect.emplace_back( fake );
```

Comment:
You can add the content of one vector to another one like this:
```suggestion
    hit_you_effect.insert( hit_you_effect.end(), rhs.hit_you_effect.begin(), rhs.hit_you_effect.end() );
```
Can be used for the other vector above as well.